app/task_routes.py

```
@@ -0,0 +1,177 @@
+        elif description_query:
+            tasks = Task.query.filter_by(description=description_query)
+        elif completed_at_query:
+            tasks = Task.query.filter_by(completed_at=completed_at_query)
+        elif is_complete_query:
+            tasks = Task.query.filter_by(is_complete=is_complete_query)
+        else:
+            tasks = Task.query.all()
+
+        for task in tasks:
+            task_object = task.to_dict()
+            tasks_response.append(task_object["task"])
+        
+        return jsonify(tasks_response), 200
+
+#PATCH route for complete task
+@task_bp.route("/<task_key>/mark_complete", methods=["PATCH"])
+def mark_task_complete_by_id(task_key):
+    
+    #validate task id
+    task = validate_task_id(task_key)
+
+
+    task.completed_at = datetime.now()
+        
+
+    db.session.commit()
+
+    #POST to Slack with SlackBot after marked complete
+    post_to_slack(task)
+
+    return jsonify(task.to_dict()), 200
+
+#PATCH route for incomplete task
+@task_bp.route("/<task_key>/mark_incomplete", methods=["PATCH"])
+def mark_task_incomplete_by_id(task_key):
+
+    #validate task id
+    task = validate_task_id(task_key)
+
+    #if task has a datetime value, changed completed_at to None
+    task.completed_at = None
+
+    db.session.commit()
+
+    #response body
+    return jsonify(task.to_dict()), 200
+
+
+#GET/READ one task by id
+@task_bp.route("/<task_key>", methods=["GET"])
+def get_one_task_by_id(task_key):
+    task = validate_task_id(task_key)
+
+    if task.goal_key:
+        return jsonify(task.dict_rel()), 200
+    
+
+    return jsonify(task.to_dict()), 200
+
+
+#REPLACE ALL OF/PUT task
+#this endpoint will update the created_at attribute wave 3
+@task_bp.route("/<task_key>", methods=["PUT"])
+def replace_task_by_id(task_key):
+    task = validate_task_id(task_key)
+    request_body = request.get_json()
+    
+    task.title = request_body["title"]
+    task.description = request_body["description"]
+    
```

Consider moving this logic into a helper in Task.

And be sure to perform the same `KeyError` handling here as you did in POST. These keys are required and should return an error message if they are left off.